fdbserver/RestoreMaster.actor.cpp

```diff
@@ -862,4 +904,67 @@ ACTOR static Future<Void> signalRestoreCompleted(Reference<RestoreMasterData> se
 	TraceEvent("FastRestore").detail("RestoreMaster", "AllRestoreCompleted");
 
 	return Void();
+}
+
+// Update the most recent time when master receives hearbeat from each loader and applier
+ACTOR static Future<Void> updateHeartbeatTime(Reference<RestoreMasterData> self) {
+	int numRoles = self->loadersInterf.size() + self->appliersInterf.size();
+	state std::map<UID, RestoreLoaderInterface>::iterator loader = self->loadersInterf.begin();
+	state std::map<UID, RestoreApplierInterface>::iterator applier = self->appliersInterf.begin();
+	state std::vector<Future<RestoreCommonReply>> fReplies(numRoles, Never()); // TODO: Reserve memory for this vector
+	state std::vector<UID> nodes;
+	state int index = 0;
+
+	// Initialize nodes only once
+	loader = self->loadersInterf.begin();
+	applier = self->appliersInterf.begin();
+	while (loader != self->loadersInterf.end()) {
+		nodes.push_back(loader->first);
+		loader++;
+	}
+	while (applier != self->appliersInterf.end()) {
+		nodes.push_back(applier->first);
+		applier++;
+	}
+
+	loop {
+		wait(delay(SERVER_KNOBS->FASTRESTORE_HEARTBEAT_DELAY));
+		loader = self->loadersInterf.begin();
+		applier = self->appliersInterf.begin();
+		index = 0;
+		std::fill(fReplies.begin(), fReplies.end(), Never());
```

Comment:
This seems redundant, since line 914 already does this.

Reply:
This is not redundant when it executes the next loop.
We need to ensure the `fReplies` are not ready at the start of the loop, because we use the its readyness to check if we get a reply.

Reply:
You are right. I missed that point.